### PR TITLE
blockchain code implementation

### PR DESCRIPTION
**Pull request by astherath**
- updated gitignore and started work on blockchain code
- more boilerplate work on block code and module refactor.
- small restructure (collapse of modules) and more documentation
- started work on implementing the linked list with lifetimes
- switched from static lifetimes to reference counters for recursive struct
- going back to lifetime parameters (nightmarish, i know)
- switched to boxes to avoid lifetimes (suffering)
- finally fixed append method and learned about borrow checker.
- refactoring and trying to pass lint
- minor lint fixes
- moved some code around for testing access and created lib for util code
- wrote small (failing tests). Changing data structures before moving on.
- totally refactored the list data structure (now working).
- added initial test suite and fixed linter errors
- added readme

## TL;DR
First round of implementation for the core block/blockchain code using a simple append-only-vector and added minimal tests/documentation.

## Implementation Details
The first data structure that came into my head when thinking "append only list" was a singly linked list where we could keep a pointer to the last node at all times.

I was wrong.

Implementing a linked list in rust is nothing short of an absolute nightmare. I thought it was just me being a beginner but then I came accross a [great article](https://rust-unofficial.github.io/too-many-lists/) that validated my experience.

The basic reason as to why it's so difficult is with the rust ownership system and references. The details are too extensive to explain in a PR but it boiled down to not being able to properly iterate or append the list without consuming the parameter (due to ownership issues).

After more thought and some refactoring, an interfaced `Vec<Block>` was the solution I ended up choosing, and it's working very well.

Even better, it's most likely that a `Vec` based implementation is going to be better for our performance in the bottom line, so that was a nice bonus.

## Other Notes
This PR also created some github actions pipelines for the code, most notable a linter (`clippy`) and a quick run of `cargo test`. The linter is set to fail on warning so make sure there's no lint errors when submitting or else you won't get a nice green check.

### CODE REVIEW

**Comment by astherath** [comment]
fixing GH checks before mergins

**Review by LazaroHurtado — COMMENTED** [review]
Awesome
